src/advanced_unsafety/invalid_values.md

```diff
@@ -0,0 +1,112 @@
+
+```rust
+enum Stuff {
+    Char(char),
+    Number(u32),
+}
+```
+
+setting the discriminant bit to something that is not the discriminant of `Char` or `Number` is invalid. Similarly, setting the discriminant bit to that for `Char` but having the value be invalid for a `char` is also invalid.
+
+### Smart pointers and NonNull
+
+Most smart pointer types like `Box<T>` and `Rc<T>` are invalid when null. Library types may achieve the same behavior using the [`NonNull<T>`] pointer type.
+
+It's also currently invalid for `Vec<T>` to have a null pointer for its buffer! `Vec<T>` uses [`NonNull<T>`] internally, and empty vectors use a pointer value equal to the alignment of `T`.
+
+
+
+### `#[repr(Rust)]` isn't stable!
+
+Note that Rust's default representation for types is not stable! What might be a valid bit pattern one day may become invalid later, unless you're only relying on things that are known to be invariant.
+
+### Invalid values for general library types
+
+In general, types may have various invalid values based on their internal representation (which may not be stable!).
+
+As a library user you may not assume anything about the representation of a library type unless it is explicitly documented as such, or if it has a public representation that is known to be stable (for example a public `#[repr(C)]` enum)
+
+## When you might end up making an invalid value
+
+
+Invalid values have a chance to crop up when you're reinterpreting a chunk of memory as a value of a different type. This can happen when calling[`mem::transmute()`] and [`mem::transmute_copy()`], or when casting a reference to a region of memory into one of a different type. The value need not be on the stack to be considered invalid: if you gin up an `&bool` that points to a bit pattern that is not a valid `bool`, that is instantly UB even if you don't read from the reference.
+
+They can also happen when receiving values over FFI where either the signature of the function is incorrect (e.g. saying an FFI function accepts `bool` when the other side thinks it accepts a `u8`), or where there are differences in notions of validity across languages.
+
+A subtle case of this comes up occasionally in FFI code due to differences in expectations between how enums are used in Rust and C.
+
+In C, it is common to use enums to represent _bitmasks_, doing something like this:
+
+```c
+typedef enum {
+    Active = 0x01;
+    Visible = 0x02;
+    Updating = 0x03;
+    Focused = 0x04;
+} NodeStatus;
+```
+
+where the value make take states like `Active | Focused | Visible`. These combined values, as well as the "no flags set" value `0` are invalid in Rust. If this type is represented as an enum in Rust ([even if it is `#[repr(C)]`][reprc-enum]!), it will be UB to accept values of this type over FFI from C. Generally in such cases it is recommended to use an integer type instead, and represent the mask values as constants.
+
+
+## Things you might see if you used invalid data
```

Comment:
I'm not quite sure on this structure just yet but I like having a section on "where you may have things go wrong" and "what can happen" ("what the optimizer may do" and also "ways to intuit that this might be the problem")